### PR TITLE
Partial theme merging fixes (for ThemeProvider) and update createTheme to use Theme/PartialTheme

### LINKED ISSUE #15280: react-theme-provider does not style Fluent UI components correctly with imported theme
### Environment Information

- **Package version(s)**: 
```
  "@fluentui/react": "^7.142.0",
  "@fluentui/react-theme-provider": "^0.14.2",
  "@uifabric/fluent-theme": "^7.1.91",
  "@uifabric/icons": "^7.3.48",
  "@uifabric/react-hooks": "^7.13.5",
```
- **Browser and OS versions**: n/a

### Please provide a reproduction of the bug in a codepen:

**codepen example below wasn't loading**

#### Creating a theme using the [Fluent UI Theme Designer](https://fabricweb.z5.web.core.windows.net/pr-deploy-site/refs/heads/master/theming-designer/index.html)

```
/src/themes.js

export const lightTheme = {
  palette: {
    themePrimary: '#148214',
    themeLighterAlt: '#f2faf2',
    themeLighter: '#cbebcb',
    themeLight: '#a2daa2',
    themeTertiary: '#58b458',
    themeSecondary: '#259125',
    themeDarkAlt: '#127512',
    themeDark: '#0f630f',
    themeDarker: '#0b490b',
    neutralLighterAlt: '#faf9f8',
    neutralLighter: '#f3f2f1',
    neutralLight: '#edebe9',
    neutralQuaternaryAlt: '#e1dfdd',
    neutralQuaternary: '#d0d0d0',
    neutralTertiaryAlt: '#c8c6c4',
    neutralTertiary: '#a19f9d',
    neutralSecondary: '#605e5c',
    neutralPrimaryAlt: '#3b3a39',
    neutralPrimary: '#323130',
    neutralDark: '#201f1e',
    black: '#000000',
    white: '#ffffff',
  }
}
```
#### Using [@fluentui/react-theme-provider](https://www.npmjs.com/package/@fluentui/react-theme-provider)

```
/src/index.jsx

...import statements...

import { ThemeProvider } from '@fluentui/react-theme-provider'; 

import { lightTheme } from './themes';

ReactDOM.render(
    <ThemeProvider theme={lightTheme}>
        <App />
    </ThemeProvider>,
  document.getElementById('root')
```

#### Import Fluent Component

```
/src/components/App.jsx

import { PrimaryButton } from '@fluentui/react';

function App() {
  return <PrimaryButton text="Standard" />;
}
```

#### Actual behavior:

Using Fluent UI components do not get styled correctly according the theme palette object provided. When using `Fabric` and `loadTheme` all components were adhering to the provided custom theme. 

Screen Shot 2020-09-28 at 10 07 27 AM

#### Expected behavior:

Fluent UI components should be styled according to exported palette object from the Fluent UI Theme Designer. I have noticed that some elements are using the theme provided but not in a complete "across the board" manner as shown in the Toggle gif below.

Screen Shot 2020-09-28 at 10 09 02 AM

In this example the [Fluent UI Toggle](https://developer.microsoft.com/en-us/fluentui#/controls/web/toggle) is filled with blue when activated. However on hover instead of going to a darker blue hue it turns green with is indicative of the theme being loaded in. The fill for this toggle when activated should be green, and when hovered a darker green should render. Note the icon should also be green and not blue

fluent-theme-toggle

### Priorities and help requested:

Are you willing to submit a PR to fix? I'm willing to help work on documentation. imo the documentation is difficult to follow and the examples are lacking completeness. 

Requested priority: High

Products/sites affected: (if applicable)

------

_Also reading the documentation, `createTheme` is imported but never used in the example of applying component-level styles._

Screen Shot 2020-09-28 at 10 00 42 AM

### PR DESCRIPTION
**Pull request by xugao**
#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000
- [x] Include a change request file using `$ yarn change`

#### Description of changes
Cherry-pick #15452 

- Update createTheme to use `Theme`/`PartialTheme` instead of `ITheme`/`IPartialTheme`
- Marked `tokens` in `Theme` as `internal` and updated its typing to be partial. This is because when `createdTheme` returns the full `Theme`. We don't necessarily have to create full `Tokens` in that case, we can just pass `tokens` through.
- Fix a bug where user provided tokens get overwritten by mapped tokens from legacy theme during `getTokens`. This is causing nested `ThemeProvider` not working for `react-button` (modified screener tests to cover this case).
- Added typing `FunctionComponent<ThemeProviderProps>` to `ThemeProvider`. This is to have cleaner public API, instead of relying of `React.forwardRef` typing.
- Fix #15280:  root cause [here](https://github.com/microsoft/fluentui/issues/15280#issuecomment-703817785)
  - Changed `mergeThemes` implementation to merge partial/full theme on top of a full theme. And use that in both `createTheme` and `useThemeProviderState`. (Note: **this is a breaking change**. however, I think it's safe to make since this is recently moved from `react-theme-provider` to `theme` and I don't think anyone but our code is using this)

#### Focus areas to test

(optional)

### CODE REVIEW

File: packages/theme/src/mergeThemes.ts [hunk 1]
```diff
@@ -1,15 +1,39 @@
...
+      partialTheme.effects,
+      partialTheme.semanticColors,
+      partialTheme.isInverted === undefined ? theme.isInverted : partialTheme.isInverted,
+    ),
+  }) as Theme;
 
-  // Correctly merge stylesheets array
-  partialTheme.stylesheets = [];
-  themes.forEach(theme => theme && theme.stylesheets && partialTheme.stylesheets?.push(...theme.stylesheets));
+  if (partialTheme.palette?.themePrimary && !partialTheme.palette?.accent) {
```
Comment (dzearing) [inline packages/theme/src/mergeThemes.ts:18]: is `accent` a value? We'd agreed on brand.

File: packages/theme/src/createTheme.ts [hunk 2]
```diff
@@ -1,57 +1,27 @@
...
 
-  const defaultFontStyles: IFontStyles = { ...DefaultFontStyles };
-
-  if (theme.defaultFontStyle) {
-    for (const fontStyle of Object.keys(defaultFontStyles) as (keyof IFontStyles)[]) {
-      defaultFontStyles[fontStyle] = merge({}, defaultFontStyles[fontStyle], theme.defaultFontStyle);
-    }
-  }
-
-  if (theme.fonts) {
```
Comment (xugao) [inline packages/theme/src/createTheme.ts:36]: do not iterate

File: packages/theme/src/mergeThemes.ts [hunk 3]
```diff
@@ -1,15 +1,39 @@
...
+  }
 
-  return partialTheme as TResult;
+  if (partialTheme.defaultFontStyle) {
+    for (const fontStyle of Object.keys(mergedTheme.fonts) as (keyof IFontStyles)[]) {
+      mergedTheme.fonts[fontStyle] = merge(mergedTheme.fonts[fontStyle], partialTheme.defaultFontStyle);
+    }
+  }
+
+  if (partialTheme.fonts) {
```
Comment (dzearing) [inline packages/theme/src/mergeThemes.ts:28]: can this be optimized?

File: packages/theme/src/mergeThemes.ts [hunk 4]
```diff
@@ -1,15 +1,39 @@
...
+      partialTheme.isInverted === undefined ? theme.isInverted : partialTheme.isInverted,
+    ),
+  }) as Theme;
 
-  // Correctly merge stylesheets array
-  partialTheme.stylesheets = [];
-  themes.forEach(theme => theme && theme.stylesheets && partialTheme.stylesheets?.push(...theme.stylesheets));
+  if (partialTheme.palette?.themePrimary && !partialTheme.palette?.accent) {
+    mergedTheme.palette.accent = partialTheme.palette.themePrimary;
+  }
```
Comment (xugao) [inline packages/theme/src/mergeThemes.ts:20]: remove `accent` palette and `accent*` semantic colors

**Review by dzearing — APPROVED** [review]
Address comments.

**Review by dzearing — APPROVED** [review]

**Pull request merged**